### PR TITLE
Reverse legend

### PR DESCRIPTION
**Pull request by Marvvxi**
## PR Summary

I added a parameter for the reverse legend option. See the following images. Taken from #24712.

without_reverse
with_reverse

## PR Checklist

**Documentation and Tests**
- [N/A] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A]API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Review by timhoffm — COMMENTED** [review]
Thanks for the contribution.

Please add a test.

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -518,7 +519,7 @@ def val_or_rc(val, rc_name):
         self._alignment = alignment
 
         # init with null renderer
-        self._init_legend_box(handles, labels, markerfirst)
+        self._init_legend_box(handles, labels, reverse, markerfirst)
```
[the remark this answers is not in the document]
Comment (Marvvxi) [reply lib/matplotlib/legend.py]: Hey thanks for your review.
I finished the changes.

File: lib/matplotlib/tests/test_legend.py [hunk 2]
```diff
@@ -294,6 +294,21 @@ def test_legend_remove():
...
+    y = [1, 2, 3], [2, 3, 4], [3, 4, 5]
+    labels = ["First label", "Second label", "Third label"]
+
+    ax.plot(x, y, label=labels)
+    leg = ax.legend(reverse=True)
+    actual_labels = []
+    for text in leg.get_texts():
+        actual_labels.append(text.get_text())
+    excepted_labels = list(reversed(labels))
+    assert actual_labels == excepted_labels
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_legend.py:309]: You can shorten this to:

```suggestion
    actual_labels = [t.get_gext() for t in leg.get_texts()]
    assert actual_labels == list(reversed(labels))
```

Additionally you should check the markers, which you can get via
```
markers = [h.get_marker() for h in leg.legend_handles]
```
I think you have to unroll the `plot` command into three separate commands to be able to specify markers explicitly (we intentionally don't support vectorization of keyword arguments in `plot`).

Please also note that `leg.legend_handles` was just added today to the main branch, so you need to rebase on `main` to be able to use that. - Let us know if you need help with rebasing.

**Review by timhoffm — COMMENTED** [review]
Thanks, this is very much in the right direction. Only one extension to the test left.

File: lib/matplotlib/legend.py [hunk 3]
```diff
@@ -431,6 +435,10 @@ def val_or_rc(val, rc_name):
                 _hand.append(handle)
         labels, handles = _lab, _hand
 
+        if reverse:
+            labels.reverse()
+            handles.reverse()
+
```
[the remark this answers is not in the document]
Comment (timhoffm) [reply lib/matplotlib/legend.py:441]: I checked that as well :smile:. We are creating new lists `labels` and `handles` just in the lines above. So yes, they always have a `.reverse()` and no overwriting is not an issue.

**Comment by Marvvxi** [comment]
Ups, sorry for the chaos. I tried something, but now my comments in my local repository behind the main comments. If I try to push, I get the message everything up-to-date. Can you help my to clean up my branch, please?

**Comment by Marvvxi** [comment]
> Thanks for implementing this! It needs a second approval to merge, but at that point would you like to try rebasing again (w/ help from folks in incubator) or would you like us to squash merge?

Thanks for approving and introduction into the development with matplotlib. I would say that the squash merge option is the better. There are some small not so important commits.

**Comment by tacaswell** [comment]
This should definitely get a whats_new entry!

I think we can sneak this in for 3.7.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Thanks @Marvvxi! To keep things simple for you and not go through additional iterations, I've done the final polishing and added a "What's new" entry as well as an additional figure comparison test that the displayed order is indeed reversed.

@story645 please re-review the additions, and squash-merge when ok. Since my additions are only documetation and a test, I figure we can merge without a third reviewer (I as a the second reviewer have contributed, but only non-code parts).
